packages/structures/src/bitfields/BitField.ts

```diff
@@ -0,0 +1,205 @@
+import type { EnumLike, RecursiveReadonlyArray } from '../utils/types';
+
+/**
+ * Data that can be resolved to give a bitfield. This can be:
+ * A bit number (this can be a number literal or a value taken from {@link BitField.Flags})
+ * A string bit number
+ * An instance of BitField
+ * An Array of BitFieldResolvable
+ */
+export type BitFieldResolvable<Flags extends string, Type extends bigint | number> =
+	| Flags
+	| Readonly<BitField<Flags, Type>>
+	| RecursiveReadonlyArray<Flags | Readonly<BitField<Flags, Type>> | Type | `${bigint}`>
+	| Type
+	| `${bigint}`;
+
+/**
+ * Data structure that makes it easy to interact with a bitfield.
+ */
+export abstract class BitField<Flags extends string, Type extends bigint | number = number> {
+	/**
+	 * Numeric bitfield flags.
+	 * <info>Defined in extension classes</info>
+	 */
+	public static Flags: EnumLike<unknown, bigint | number> = {};
```

Comment:
What was the thought process behind `DefaultBit` being typed like this (I guess Flags you can sort of justify it, though I'd probably go `Type | string` there too) despite the `Type` generic? I do see the way we handle it in `resolve` but I feel like it's begging for trouble? Imagine ending up with this by refactoring

```ts
class Something extends BitField<Flags, bigint> { // just changed to bigint
  public static override Flags = whatever; // contains a bigint that would be too big for number representation
  public static DefaultBit = 123; // uh-oh! forgot to change this. everything breaks
}
```

Reply:
static properties can't use the generic, as it's only present on instances, not the class itself.

Reply:
Right. That's quite annoying. Anyone else have ideas on how to improve the safety here?

Reply:
Maybe inspire from [@sapphire/bitfield](https://www.npmjs.com/package/@sapphire/bitfield) and make the enum be passed in? But then we'd basically have a factory of bitfields 🫠...

Reply:
I'm open for any way to do this here. The current implementation is just a 1:1 port of current mainlib BitField and its manual typings.